### PR TITLE
support backward return None, when corresponding input tensor without gradient

### PR DESCRIPTION
**Pull request by hbwx24**
### PR types

 Function optimization
### PR changes

OPs
### Describe

当pylayer_op的forward输入tensor的stop_gradient=True时，backward中对应的梯度可以返回None。
如果对应的tensor的stop_gradient=False，不能返回None。

```
class tanh(PyLayer):
      @staticmethod
      def forward(ctx, x1, x2, func1, func2=paddle.square):
          ctx.func = func2
          y1 = func1(x1)
          y2 = func1(x2)
          ctx.save_for_backward(y1, y2)
          return y1, y2

      @staticmethod
      def backward(ctx, dy1, dy2):
          y1, y2 = ctx.saved_tensor()
          re1 = dy1 * (1 - ctx.func(y1))
          re2 = dy2 * (1 - paddle.square(y2))
          return re1, None

input1 = paddle.randn([2, 3]).astype("float64")
input2 = input1.detach().clone()
input3 = input1.detach().clone()
input4 = input1.detach().clone()
input1.stop_gradient = False
input2.stop_gradient = False
input3.stop_gradient = True
input4.stop_gradient = True
z = tanh.apply(input1, input3, paddle.tanh, paddle.square)
z = z[0] + z[1]
z.mean().backward()

z2 = paddle.tanh(input2) + paddle.tanh(input4)
z2.mean().backward()

print(np.max(np.abs((input1.grad - input2.grad))) < 1e-10)
```

### CODE REVIEW

File: paddle/fluid/operators/py_layer_op.cc [hunk 1]
```diff
@@ -60,34 +60,41 @@ void RunPyObject(py::object *py_object,
...
+        } else {
+          PADDLE_THROW(platform::errors::Unimplemented(
+              "The %dth input tensor of forward needs gradient and the "
+              "corresponding gradient cannot be None.",
+              i));
+        }
+      }
+    }
+  } else {
+    if ((*outs)[0] != nullptr) {
```
Comment (chenwhql) [inline paddle/fluid/operators/py_layer_op.cc:89]: 这里如果是nullptr的话需要报错吗

File: paddle/fluid/operators/py_layer_op.cc [hunk 2]
```diff
@@ -60,34 +60,41 @@ void RunPyObject(py::object *py_object,
...
+              py_result.cast<std::shared_ptr<imperative::VarBase>>();
+          *((*outs)[0]) = result_var->Var();
         } catch (py::cast_error &) {
           PADDLE_THROW(platform::errors::Unimplemented(
               "The output of `PyLayer.backward` should be `Tensor`."));
         }
       } else {
         PADDLE_THROW(platform::errors::Unimplemented(
-            "The output of `PyLayer.backward` can not be `None`."));
+            "The input tensor of forward needs gradient, so the output of "
```
Comment (chenwhql) [inline paddle/fluid/operators/py_layer_op.cc:101]: 这里在什么地方判断前向输入的stop_gradient信息的

File: paddle/fluid/operators/py_layer_op.cc (hunk 1 above)
Comment (hbwx24) [reply paddle/fluid/operators/py_layer_op.cc:89]: 完善了报错，如果是nullptr，backward必须返回None，否则报错。

File: paddle/fluid/operators/py_layer_op.cc (hunk 2 above)
Comment (hbwx24) [reply paddle/fluid/operators/py_layer_op.cc:101]: grad op的第i个输出(*outs)[i] 为nullptr，由此判断第i个输入是不需要梯度的。

**Review by ForFishes — APPROVED** [review]
LGTM

**Review by chenwhql — APPROVED** [review]
